vpr/src/route/connection_box_lookahead_map.cpp

```diff
@@ -213,6 +228,404 @@ class CostMap {
 
 static CostMap g_cost_map;
 
+// Search grid used for sampling segments from graph.
+//
+// SearchGrid divides the graph based on segment type, and within each
+// segment type defines a NxN uniform search grid, where N is kTargetGridSize.
+//
+// SearchGrid needs to be initialized with the setup and add_node methods.
+//
+// Once initialized, next_node can be invoked to determine next rr node to
+// expand.
+//
+// During expansion, visit_entry and expanded should be called to indicate
+// that an expansion found data relevant to the search.
+class SearchGrid {
+    // Representation of one search area of the grid.
+    struct GridEntry {
+        GridEntry()
+            : samples(0)
+            , bounds(vtr::Rect<int>(-1, -1, -1, -1)) {}
+
+        // Number of samples take from this section of the grid.
+        int samples;
+
+        // Geometric boundry of this section of the grid.
+        vtr::Rect<int> bounds;
+
+        // rr nodes that belong to this section of the grid.
+        std::vector<int> nodes;
+    };
+
+    // Represent of one segment (element) type within the grid.
+    struct ElementDefinition {
+        // Bounding box that contains all instances of a particular segment type.
+        t_bb box;
+
+        // Stride is choosen such that dim_size(0)*x_stride and
+        // dim_size(1)*y_stride encompass box (e.g. the grid includes all
+        // elements).
+        size_t x_stride;
+        size_t y_stride;
+
+        // Search grid for this element.
+        vtr::NdMatrix<GridEntry, 2> grid;
+    };
+
+  public:
+    // Initial search grid with bound boxes of each search instance, and total
+    // number of nodes that could be expanded.
+    void setup(const std::vector<t_bb>& grid_elements, size_t num_nodes) {
+        expanded_.resize(num_nodes, false);
+
+        constexpr int kTargetGridSize = 3;
+        elements_.resize(grid_elements.size());
+
+        for (size_t i = 0; i < grid_elements.size(); ++i) {
+            elements_[i].box = grid_elements[i];
+            if (grid_elements[i].xmin > grid_elements[i].xmax) {
+                continue;
+            }
+
+            VTR_ASSERT(grid_elements[i].ymin <= grid_elements[i].ymax);
+
+            size_t x_dim = std::min(kTargetGridSize, grid_elements[i].xmax - grid_elements[i].xmin + 1);
+            size_t y_dim = std::min(kTargetGridSize, grid_elements[i].ymax - grid_elements[i].ymin + 1);
```

Comment:
If you use `vtr::Rect` you can use the `width` and `height` methods.

Reply:
See above